src/Database/Statement/Statement.php

```diff
@@ -0,0 +1,328 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         5.0.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Database\Statement;
+
+use Cake\Database\DriverInterface;
+use Cake\Database\FieldTypeConverter;
+use Cake\Database\Log\LoggedQuery;
+use Cake\Database\StatementInterface;
+use Cake\Database\TypeConverterTrait;
+use Cake\Database\TypeMap;
+use InvalidArgumentException;
+use PDO;
+use PDOException;
+use PDOStatement;
+use Psr\Log\LoggerInterface;
+
+class Statement implements StatementInterface
+{
+    use TypeConverterTrait {
+        cast as protected;
+        matchTypes as protected;
+    }
+
+    /**
+     * @var array<string, int>
+     */
+    protected const MODE_NAME_MAP = [
+        self::FETCH_TYPE_ASSOC => PDO::FETCH_ASSOC,
+        self::FETCH_TYPE_NUM => PDO::FETCH_NUM,
+        self::FETCH_TYPE_OBJ => PDO::FETCH_OBJ,
+    ];
+
+    /**
+     * @var \Cake\Database\DriverInterface
+     */
+    protected DriverInterface $_driver;
+
+    /**
+     * @var \PDOStatement
+     */
+    protected PDOStatement $statement;
+
+    /**
+     * @var \Cake\Database\FieldTypeConverter|null
+     */
+    protected ?FieldTypeConverter $typeConverter;
+
+    /**
+     * @var \Psr\Log\LoggerInterface|null
+     */
+    protected ?LoggerInterface $logger = null;
+
+    /**
+     * Cached bound parameters used for logging
+     *
+     * @var array<mixed>
+     */
+    protected array $params = [];
+
+    /**
+     * @var float
+     */
+    protected float $took = 0.0;
+
+    /**
+     * @param \PDOStatement $statement PDO statement
+     * @param \Cake\Database\DriverInterface $driver Database driver
+     * @param \Cake\Database\TypeMap|null $typeMap Results type map
+     */
+    public function __construct(
+        PDOStatement $statement,
+        DriverInterface $driver,
+        ?TypeMap $typeMap = null,
+    ) {
+        $this->_driver = $driver;
+        $this->statement = $statement;
+        $this->typeConverter = $typeMap !== null ? new FieldTypeConverter($typeMap, $driver) : null;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function setLogger(LoggerInterface $logger)
+    {
+        $this->logger = $logger;
```

Comment:
Instead of setting the logger on each statement, could we get the logger from the driver? It would require pushing the logger down from `Connection` into the `Driver` but it would remove some complexity here.

If you'd like I can take a swing at doing that change.

Reply:
What is the difference between Connection and Driver long-term? Do we want Driver to maintain state?